[llvm-objcopy] Add --gap-fill and --pad-to options

quic-akaryaki opened the pull request:
`--gap-fill <value>` fills the gaps between sections with a specified 8-bit value, instead of zero.
`--pad-to <address>` pads the output binary up to the specified load address, using the 8-bit value from `--gap-fill` or zero.

These options are only supported for ELF inputs and binary outputs.

quic-akaryaki committed:
[llvm-objcopy] Add --gap-fill and --pad-to options

`--gap-fill <value>` fills the gaps between sections with a specified
8-bit value, instead of zero.
`--pad-to <address>` pads the output binary up to the specified load
address, using the 8-bit value from `--gap-fill` or zero.

These options are only supported for ELF inputs and binary outputs.

quic-akaryaki commented:
This is an implementation of section padding with a more limited scope than was attempted in https://reviews.llvm.org/D67689.
Only binary outputs are supported, but ELF and IHEX are not. The main motivation for taking this approach was to support the clear use case of building padded ROM images without figuring all the murky details of the general ELF case.
After reviewing https://reviews.llvm.org/D67689, it seems the main problem was with how the section content is updated for padding. The SectionBase::OriginalData was modified, which is incorrect. However, I think the bigger problem is that there are multiple section types that represent their data differently (SectionBase::OriginalData, OwnedDataSection::Data, compressed sections, etc). In binutils, sections are represented uniformly and resizing them in one location is sufficient. I don't know if a similar simple solution is possible in llvm-objdump. Therefore, I think it would make sense to handle padding in the output stage.
I think which approach is better going forward ultimately depends on whether section padding is a valuable use case for ELF and IHEX outputs. I am not aware of such use cases, but suggestions are welcome. Talking about ROM images, it is possible in some platforms to directly flash an ihex. However, since ihex contains addresses, it supports gaps naturally and does not have to store all the padding bytes. During the flashing, the gaps are skipped.

MaskRay commented:
(It seems that pr-subscribers-llvm-binary-utilities cannot be added as a reviewer)

quic-akaryaki commented:
I'm not familiar with the new github PR flow. I cannot add reviewers myself and there is no automatic assignment/recommendation.

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/gap-fill-order.test:
```diff
@@ -0,0 +1,34 @@
+# RUN: yaml2obj %s >%t
```
Nit: most of our newer tests use the `-o` option instead of output redirection (which has potential issues when copy-pasted into certain terminals, e.g. PowerShell).

quic-akaryaki replied:
I change it to use the `-o` option now.

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/gap-fill-order.test:
```diff
@@ -0,0 +1,34 @@
+# RUN: yaml2obj %s >%t
+
+# Test gap fill with all allocatable output sections
```
Nit: New llvm-objcopy tests use `##` for the comments, to help them stand out from RUN and CHECK lines. 

Also, this sentence needs a trailing "." - this applies in several other places in the tests too.

This comment doesn't really help me understand the purpose of this test. The test name is `gap-fill-order.test` which implies that it has something to do with the order of something, but it's not clear what that is.

quic-akaryaki replied:
Fixed the comments and some punctuation in comments and added a comment.

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/gap-fill-order.test:
```diff
@@ -0,0 +1,34 @@
+  Class:           ELFCLASS32
+  Data:            ELFDATA2LSB
+  Type:            ET_EXEC
+  Machine:         EM_HEXAGON
```
Hexagon is an unusual Machine type. Is it possible to omit the machine type completely in this case? If not, can it be x86_64?

quic-akaryaki replied:
This is true, I changed it to x86-64.

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/gap-fill-order.test:
```diff
@@ -0,0 +1,34 @@
+    AddressAlign:    0x0001
+    Content:         'AABBCCDD'
+...
+
```
Nit: files should end with exactly one \n. It looks like this file might have a double one.

quic-akaryaki replied:
Fixed.

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/gap-fill.test:
```diff
@@ -0,0 +1,149 @@
+# EMPTY: no input file specified
+
+# RUN: not llvm-objcopy -O binary --gap-fill= %t 2>&1 | FileCheck %s --check-prefix=BAD-FORMAT
+# BAD-FORMAT: bad number for --gap-fill:
```
This error message as checked doesn't explain how the number is bad. In particular, I think it's important to distinguish the "not 8-bits" from the "not a valid number" cases. Also, I think it would be worth testing hex and decimal cases.

quic-akaryaki replied:
And I have added test cases with decimal numbers.

jh7370 replied:
Our error messages don't need to exactly match GNU's. If we can provide more useful messages, we should do so.

jh7370 commented on llvm/docs/CommandGuide/llvm-objcopy.rst:
```diff
@@ -324,6 +324,11 @@ them.
 
  Extract the named partition from the output.
 
+.. option:: --gap-fill <value>
```
In addition to the CommandGuide update, you should update the release notes.

jh7370 commented on llvm/lib/ObjCopy/ELF/ELFObject.cpp:
```diff
@@ -2635,9 +2635,36 @@ template <class ELFT> Error ELFWriter<ELFT>::finalize() {
+
+  assert(LoadableSections.front()->Offset == 0);
+
+  for (unsigned i = 0; i != LoadableSections.size(); ++i) {
```
`size_t` not unsigned to match the return type of `size`

quic-akaryaki replied:
Changed this to std::size_t.

jh7370 commented on llvm/lib/ObjCopy/ELF/ELFObject.cpp:
```diff
@@ -2635,9 +2635,36 @@ template <class ELFT> Error ELFWriter<ELFT>::finalize() {
-  for (const SectionBase &Sec : Obj.allocSections())
+  SmallVector<const SectionBase *, 30> LoadableSections;
+  for (const SectionBase &Sec : Obj.allocSections()) {
+    if ((Sec.Flags & SectionFlag::SecLoad) && Sec.Type != SHT_NOBITS)
```
Skipping NOBITS sections seems a fairly dramatic departure from the previous code. It may make sense, but I'd like to hear your explanation.

quic-akaryaki replied:
This is interesting. The first part of this condition was wrong, I don't remember why I added it, so I removed it now. However, the NOBITS check should be there. NOBITS sections are not considered when assigning offsets and determining the size of the output buffer in `BinaryWriter::finalize()`. They are also not written to the output in `SectionWriter::visit()`. Another question is whether NOBITS sections should be filled with a pattern, but GNU objcopy does this, which makes sense to me.

jh7370 replied:
> Another question is whether NOBITS sections should be filled with a pattern, but GNU objcopy does this, which makes sense to me.

FWIW, the ELF spec explicitly states that `PT_LOAD` with `p_memsz > p_filesz` is padded with zeros (this is the case for a .bss section at a the end of a segment). It would be weird for other output formats to define a different system for initializing NOBITS data.

Anyway, no concrete concerns here, if @MaskRay is happy (like I said, I don't know much about the details of binary output).

jh7370 commented on llvm/lib/ObjCopy/ELF/ELFObject.cpp:
```diff
@@ -2635,9 +2635,36 @@ template <class ELFT> Error ELFWriter<ELFT>::finalize() {
+                             ? LoadableSections[i + 1]->Offset
+                             : Buf->getBufferSize();
+    assert(PadOffset <= Buf->getBufferSize());
+    if (Sec.Offset + Sec.Size < PadOffset)
```
This if check feels a bit redundant?

quic-akaryaki replied:
I converted it to an assert. It is redundant in the case sections are sorted and assigned offsets correctly, which happens in another location.

jh7370 commented on llvm/lib/ObjCopy/ELF/ELFObject.h:
```diff
@@ -357,6 +357,8 @@ template <class ELFT> class ELFWriter : public Writer {
 class BinaryWriter : public Writer {
 private:
+  const uint8_t GapFill;
+  const std::optional<uint64_t> PadTo;
```
The `optional` is a bit unnecessary, right? Couldn't you just default `PadTo` to 0?

jh7370 replied:
It looks like this is still `optional`?

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/gap-fill.test:
```diff
@@ -0,0 +1,149 @@
+# RUN: yaml2obj %s > %t
+
+## Verify section headers before we perform several testings.
```
"tests" rather than "testings"

What is this actually supposed to be verifying though? You're using yaml2obj to create your input, so you should have full control over the output. You also aren't using this as a comparison to the later checks in the file (since those aren't checking section headers).

quic-akaryaki replied:
I cleaned up this and other comments in this file.

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/gap-fill.test:
```diff
@@ -0,0 +1,149 @@
+# NOT-BINARY: error: '--gap-fill' is only supported for binary output
+
+# RUN: not llvm-objcopy -O binary --gap-fill %t 2>&1 | FileCheck %s --check-prefix=EMPTY
+# EMPTY: no input file specified
```
Please include the "error:" prefix, to match other error checks. Same elsewhere.

quic-akaryaki replied:
Done.

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/gap-fill.test:
```diff
@@ -0,0 +1,149 @@
+# RUN: not llvm-objcopy -O binary --gap-fill=0x1122 %t %t-val16 2>&1 | FileCheck %s --check-prefix=BAD-INPUT3
+# BAD-INPUT3: bad number for --gap-fill: 0x1122
+
+# Test no gap fill gap fill with all allocatable output sections
```
Comment feels a little mangled?

quic-akaryaki replied:
Just a little.

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/gap-fill.test:
```diff
@@ -0,0 +1,149 @@
+
+# Test no gap fill gap fill with all allocatable output sections
+# RUN: llvm-objcopy -O binary %t %t-default
+# RUN: od -v -Ax -t x1 %t-default | FileCheck %s --check-prefix=DEFAULT
```
Should these have --match-full-lines, to show that there's no more output after the last section?

quic-akaryaki replied:
Added those for hexdump comparison.

jh7370 reviewed (commented):
Your test cases don't appear to involve any program headers. It's my understanding that `-O binary` is usually related to loadable programs, which have program headers. Indeed, at least one of your test cases talks about segments, but doesn't use them at all! Does your test case need updating?

You should probably update ConfigManager.cpp to indicate that the option isn't supported on other platforms.

quic-akaryaki committed:
Address PR feedback.

Change-Id: I7d57e3e87bd0393bbfe1b664228fe3bb0154d244

quic-akaryaki commented:
> Your test cases don't appear to involve any program headers. It's my understanding that `-O binary` is usually related to loadable programs, which have program headers. Indeed, at least one of your test cases talks about segments, but doesn't use them at all! Does your test case need updating?
> 
> You should probably update ConfigManager.cpp to indicate that the option isn't supported on other platforms.

objcopy does not use program headers for binary output. It only uses load addresses from section headers. I think adding them to the test would be redundant.
I have updated ConfigManager.cpp to add the new options to the list of unsupported options.

jh7370 commented on llvm/lib/ObjCopy/ELF/ELFObject.cpp:
```diff
@@ -2651,7 +2651,7 @@ Error BinaryWriter::write() {
   assert(LoadableSections.front()->Offset == 0);
 
-  for (unsigned i = 0; i != LoadableSections.size(); ++i) {
+  for (std::size_t i = 0; i != LoadableSections.size(); ++i) {
```
Nit: `size_t` seems to be the norm rather than `std::size_t` (in my current checkout, there are less than 500 lines referencing `std::size_t` versus over 12500 referencing `size_t` (so presumably ~12000 that actually use `size_t` and not `std::size_t`).

quic-akaryaki replied:
Done.

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/gap-fill-order.test:
```diff
@@ -1,16 +1,19 @@
-# RUN: yaml2obj %s >%t
+# RUN: yaml2obj %s -o %t
+
+## In this test, output sections are defined out of
+## order in respect to their load addresses. Verify
+## that gaps are still correctly filled.
```
Nit: unnecessarily early line wrapping (there's no strict coding standard for lit test files, but 80 seems like a reasonable value to me, to match other code).

quic-akaryaki replied:
Done.

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/gap-fill-order.test:
```diff
@@ -1,16 +1,19 @@
-# RUN: yaml2obj %s >%t
+# RUN: yaml2obj %s -o %t
```
I should probably have mentioned this earlier: in llvm-objcopy, where we have edge case behaviours to test for a given feature, we usually fold them into a single test file (i.e. I'd add the body of this test to `gap-fill.test`).

If it's necessary to have different YAML docs, simply use the `--docnum` option to specify which one to use.

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/gap-fill.test:
```diff
@@ -1,39 +1,32 @@
-# ORG-SHDR:  [ 4] .nobit_tbss       NOBITS          0000000000000180 000058 000018 00 WAT  0   0  8
-# ORG-SHDR:  [ 5] .foo              PROGBITS        0000000000000184 00005c 000004 00  WA  0   0  1
-# ORG-SHDR:  [ 6] .nobit_bss        NOBITS          000000000000018a 000060 000008 00  WA  0   0  1
+## This test is partially based on one from D67689.
```
This reference to "D67689" is going to get archaic very quickly given we've moved to GitHub PRs and references to DXXXXX will be lost on newer contributors. If you feel it's important, you can add a note in the commit message (but please use the full URL, so that it can be followed).

quic-akaryaki replied:
I have removed the reference.

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/gap-fill.test:
```diff
@@ -1,39 +1,32 @@
+# BAD-INPUT3: error: --gap-fill: bad number: ff
 
-# Test no gap fill gap fill with all allocatable output sections
+# RUN: llvm-objcopy -O binary --gap-fill=0x1122 %t %t-val16 2>&1 | FileCheck %s --check-prefix=TRUNCATED
```
This test should probably also show what the actual behaviour ends up being (i.e. that "0x22" is actually used).

quic-akaryaki replied:
Done.

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/gap-fill.test:
```diff
@@ -1,39 +1,32 @@
 # RUN: not llvm-objcopy --gap-fill 1 %t 2>&1 | FileCheck %s --check-prefix=NOT-BINARY
 # NOT-BINARY: error: '--gap-fill' is only supported for binary output
 
 # RUN: not llvm-objcopy -O binary --gap-fill %t 2>&1 | FileCheck %s --check-prefix=EMPTY
```
This case actually shows that `--gap-fill <something else>` treats the `<something else>` as the value. Probably you should have a test case for `--gap-fill` (without a value) being the last argument, e.g. `not llvm-objcopy -O %t --gap-fill`.

jh7370 replied:
> if the last option in the command line has no value, it is silently ignored.

This was the case that surprised me. I feel like this is a bug in the command-line option parser (it's more generic than for llvm-objcopy). Could you file a bug for this, please?

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/gap-fill.test:
```diff
@@ -1,39 +1,32 @@
 
 # RUN: not llvm-objcopy -O binary --gap-fill=x %t 2>&1 | FileCheck %s --check-prefix=BAD-INPUT
-# BAD-INPUT: bad number for --gap-fill: x
+# BAD-INPUT: error: --gap-fill: bad number: x
```
Perhaps also an `0x` case?

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/gap-fill.test:
```diff
@@ -44,10 +37,10 @@
 # DEFAULT-NEXT: 000086
 
-# Test gap fill with all allocatable output sections
+## Test gap fill with all allocatable output sections
```
```suggestion
## Test gap fill with all allocatable output sections.
```
Nit.

quic-akaryaki replied:
Done.

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/pad-to.test:
```diff
@@ -1,46 +1,56 @@
+## Save the baseline, not padded output.
+# RUN: llvm-objcopy -O binary %t %t.bin
+
+## Pad to a address smaller than the binary size.
```
```suggestion
## Pad to an address smaller than the binary size.
```
Nit.

quic-akaryaki replied:
Done.

jh7370 reviewed (commented):
I'll probably have more comments at a later date, but I'll let you update this PR to address my comments before making them.

quic-akaryaki committed:
Addressing more feedback.

Change-Id: Ic56f43f2ef4818381d98a062d1764ec127ea646e

quic-akaryaki committed:
Merge branch 'llvm:main' into gap-fill-pad-to-binary

jh7370 commented on llvm/lib/ObjCopy/ELF/ELFObject.cpp:
```diff
@@ -2635,9 +2635,36 @@ template <class ELFT> Error ELFWriter<ELFT>::finalize() {
+  SmallVector<const SectionBase *, 30> LoadableSections;
+  for (const SectionBase &Sec : Obj.allocSections()) {
+    if (Sec.Type != SHT_NOBITS)
+      LoadableSections.push_back(&Sec);
```
NOBITS sections are loadable, so I don't think this variable name makes sense.

quic-akaryaki replied:
I renamed this variable to BitsSections, maybe not the best name, but accurate.

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/gap-fill.test:
```diff
@@ -0,0 +1,198 @@
+    EntSize:         0x0001
+    Content:         4743433A
+
+## In this test, output sections are defined out of in respect to their load
```
What do you mean by "defined out of". Are you missing a word or two?

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/pad-to.test:
```diff
@@ -0,0 +1,94 @@
+# DEFAULT-NEXT: 000010 77 88 99 aa 00 00 00 00
+# DEFAULT-NEXT: 000018
+
+## Use a decimal number of padding address, verify it's not misunderstood.
```
```suggestion
## Use a decimal number for the padding address and verify it's not misunderstood.
```

quic-akaryaki replied:
Done.

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/pad-to.test:
```diff
@@ -0,0 +1,94 @@
+# FILL-NEXT: 000010 77 88 99 aa e9 e9 e9 e9
+# FILL-NEXT: 000018
+
+## Test gap fill with a section removed.
```
This test file is about `--pad-to`. This comment probably needs expanding to better explain the test case.

quic-akaryaki replied:
Done.

jh7370 commented on llvm/tools/llvm-objcopy/ObjcopyOptions.cpp:
```diff
@@ -738,6 +738,37 @@ objcopy::parseObjcopyOptions(ArrayRef<const char *> RawArgsArr,
+                   << llvm::utohexstr(ByteVal, true) << '\n';
+    Config.GapFill = ByteVal;
+  } else
+    Config.GapFill = 0; // The value of zero is equivalent to no fill.
```
The value of `GapFill` is initialized to 0 already. Do you really need this? Same below for `PadTo`.

quic-akaryaki replied:
Done.

jh7370 commented on llvm/tools/llvm-objcopy/ObjcopyOptions.cpp:
```diff
@@ -738,6 +738,37 @@ objcopy::parseObjcopyOptions(ArrayRef<const char *> RawArgsArr,
+                               A->getValue());
+    uint8_t ByteVal = Val.get();
+    if (ByteVal != Val.get())
+      llvm::errs() << "warning: truncating gap-fill from 0x"
+                   << llvm::utohexstr(Val.get(), true) << " to 0x"
+                   << llvm::utohexstr(ByteVal, true) << '\n';
```
Don't print direct to `llvm::errs`. This won't follow the proper formatting style for warnings produced by tools (see what other tools do). llvm-objcopy.cpp already has a `reportWarning` method. It seems reasonable to declare that in a header file so that you can also use it here.

quic-akaryaki replied:
Done.

quic-akaryaki committed:
Another round of review fixes.

Use reportWarning(), some renaming and rewording and other miscellaneous
changes.

quic-akaryaki commented:
I created another PR related to missing argument values: https://github.com/llvm/llvm-project/pull/70710.
It is a problem in objdump as the LLVM args parser detects a missing value, but objdump (and many other tools) ignores it.
That will be a breaking change.

quic-akaryaki committed:
Fill a missing word in a test comment.

quic-akaryaki committed:
Merge branch 'llvm:main' into gap-fill-pad-to-binary

jh7370 commented on llvm/lib/ObjCopy/ELF/ELFObject.cpp:
```diff
@@ -2636,30 +2636,30 @@ template <class ELFT> Error ELFWriter<ELFT>::finalize() {
 
 Error BinaryWriter::write() {
-  SmallVector<const SectionBase *, 30> LoadableSections;
+  SmallVector<const SectionBase *, 30> BitsSections;
```
Without knowing the code around it, I wouldn't know what `BitsSections` means. How about `SectionsToWrite`?

quic-akaryaki replied:
Done.

jh7370 commented on llvm/tools/llvm-objcopy/llvm-objcopy.cpp:
```diff
@@ -67,12 +67,18 @@ using namespace llvm::object;
+namespace llvm {
+namespace objcopy {
+
+ErrorSuccess reportWarning(Error E) {
```
Please follow the guideline described here: https://llvm.org/docs/CodingStandards.html#use-namespace-qualifiers-to-implement-previously-declared-functions

quic-akaryaki replied:
Done.

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/gap-fill.test:
```diff
@@ -163,8 +163,8 @@ Sections:
 
-## In this test, output sections are defined out of in respect to their load
-## addresses. Verify that gaps are still correctly filled.
+## In this test, output sections are defined out of order in respect to their
```
```suggestion
## In this test, output sections are defined out of order with respect to their
```
Nit: the phrase is "with respect to".

jh7370 reviewed (commented):
Basically looks good. Just some minor nits now.

quic-akaryaki committed:
Some more aesthetics.

jh7370 reviewed (approved):
I'd like @MaskRay to give this a once-over before merging this, but otherwise LGTM.

quic-akaryaki committed:
Merge remote-tracking branch 'fork/main' into HEAD

MaskRay reviewed (commented):
Sorry for the delay..

MaskRay commented on llvm/tools/llvm-objcopy/ObjcopyOpts.td:
```diff
@@ -230,3 +230,15 @@ defm add_symbol
+defm pad_to
+    : Eq<"pad-to", "Pad the output to the load address <address>, using a value "
+                   "of zero or the value specified by :option:`--gap-fill`"
+                   "This option is only supported for ELF inputs and binary outputs.">,
```
Help message typically omit the trailing period

quic-akaryaki replied:
Done. I also removed formatting, not sure why I added it here.

MaskRay commented on llvm/tools/llvm-objcopy/ObjcopyOpts.td:
```diff
@@ -230,3 +230,15 @@ defm add_symbol
+defm pad_to
+    : Eq<"pad-to", "Pad the output to the load address <address>, using a value "
+                   "of zero or the value specified by :option:`--gap-fill`"
+                   "This option is only supported for ELF inputs and binary outputs.">,
```
Missing `.` before `This option ...`?

quic-akaryaki replied:
Done.

quic-akaryaki committed:
Merge branch 'llvm:main' into gap-fill-pad-to-binary

quic-akaryaki committed:
Lastest review.

Change-Id: I3c2b19888c9e4d397a191861136c48b5715155c6

quic-akaryaki commented on llvm/test/tools/llvm-objcopy/ELF/gap-fill.test:
```diff
@@ -0,0 +1,198 @@
+    Type:            SHT_PROGBITS
+    Flags:           [ SHF_ALLOC ]
+    Address:         0x0102
+    AddressAlign:    0x0001
```
Done.

MaskRay replied:
Not done. There are still many `AddressAlign: 0x0001`

quic-akaryaki commented on llvm/test/tools/llvm-objcopy/ELF/gap-fill.test:
```diff
@@ -0,0 +1,198 @@
+## In this test, output sections are defined out of order with respect to their
+## load addresses. Verify that gaps are still correctly filled.
+
+# RUN: yaml2obj --docnum=2 %s -o %t
```
Done.

quic-akaryaki commented on llvm/tools/llvm-objcopy/ObjcopyOptions.cpp:
```diff
@@ -738,6 +739,36 @@ objcopy::parseObjcopyOptions(ArrayRef<const char *> RawArgsArr,
+                               A->getValue());
+    uint8_t ByteVal = Val.get();
+    if (ByteVal != Val.get())
+      if (Error E = reportWarning(llvm::createStringError(
```
Done. I left the changes to reportWarning as they may be useful in the future (and its less work not to undo them).

MaskRay replied:
> jh7370: I think you should undo the reportWarning changes: there's no guarantee anybody will need it in the future, so you shouldn't change it now.

Agreed. We should remove the `reportWarning` definition.

quic-akaryaki committed:
Format.

Change-Id: I44945026180d959b9616d70b910321cfc8476eff

quic-akaryaki committed:
Failing test fixed.

Change-Id: Icd803193dd02f63959b56a6f9e404c4751cc933b

jh7370 commented on llvm/tools/llvm-objcopy/ObjcopyOpts.td:
```diff
@@ -234,11 +234,11 @@ defm update_section
     : Eq<"gap-fill", "Fill the gaps between sections with <value> instead of zero. "
                      "<value> must be an unsigned 8-bit integer. "
-                     "This option is only supported for ELF inputs and binary outputs.">,
+                     "This option is only supported for the ELF input and binary output">,
```
I don't think you want "the" before "ELF".

quic-akaryaki replied:
Removed.

jh7370 commented on llvm/tools/llvm-objcopy/ObjcopyOpts.td:
```diff
@@ -234,11 +234,11 @@ defm update_section
-                   "This option is only supported for ELF inputs and binary outputs.">,
+    : Eq<"pad-to", "Pad the output up to the load address <address>, using a value "
+                   "of zero or the value specified by the --gap-fill option. "
+                   "This option is only supported for the ELF input and binary output">,
```
Ditto: no "the" before "ELF".

quic-akaryaki replied:
Removed.

jh7370 reviewed (commented):
I think you should undo the `reportWarning` changes: there's no guarantee anybody will need it in the future, so you shouldn't change it now.

quic-akaryaki committed:
Remove remaining AddressAlign from the test and undo reportWarning.

quic-akaryaki committed:
Remove thes.

jansvoboda11 commented:
The new tests started failing on Apple build bot: https://green.lab.llvm.org/green/job/clang-stage1-cmake-RA-incremental/38908/
Can you please fix these or revert the commit?

quic-akaryaki commented:
Thanks for letting me know, I'm preparing a fix. By the way, is there a way to run a buildbot manually?

jansvoboda11 commented:
> Thanks for letting me know, I'm preparing a fix. By the way, is there a way to run a buildbot manually?

Thanks! Not that I know of.

quic-akaryaki commented:
Trying a fix in #75631.
